### PR TITLE
Structure MVC architecture

### LINKED ISSUE #92: Recent contributors - is it by repository?
I went to https://code.publiclab.org/#r=image-sequencer, to see recent contributors for that project, and compared it to:

https://github.com/publiclab/image-sequencer/pulse/monthly
https://github.com/publiclab/image-sequencer/pulse/weekly

I'm not sure if it's pulling recent contributors for all of plots, though? 

image

```
@pabisovsky, @ajzliu, @Rishabh570, @radheyshyamjangid, @aashna27, @jywarren, @Radhikadua123, @icarito, @moogwol, @dependabot[bot], @shubhscoder, @cesswairimu, @samyuravikumar, @slasher17, @CoderJolly, @mrtmrtmlck, @junhouse, @narnt, @hydro-m, @kevinzluo, @romanrodriguez, @love-opensource
```

Just checking! @Rishabh570 this is such a great feature - i was going to ask recent contributors if they'd like to take up some new bugs and feature requests. It's great to be able to have this level of understanding of a growing community 😄 Your work is really helping us grow!

### PR DESCRIPTION
**Pull request by Rishabh570**
Fixes #92 

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts
* [x] PR is descriptively titled
* [x] PR body includes `fixes #0000`-style reference to original issue #
* [ ] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software

We have a loose schedule of reviewing and pulling in changes every Tuesday and Friday, and publishing changes on Fridays. 

Thanks!

### CODE REVIEW

**Comment by Rishabh570** [comment]
@jywarren Here is the WIP PR for it :)

We use `github-api-simple` for requesting to Github API, only one `fetch` goes for getting all repos (as an aside, currently I'm not storing all the repos in the database because it's stretching the code length...but I'll figure it out). 
I am thinking of storing the quota refill time so that we can make a fresh request every time the quota gets refilled...but need to figure out how to trigger a request dynamically...

For recent contributors, I'm thinking of fetching _recent-contributors-of-last-month_ and then playing with that data to filter out _recent-contributors-last-week_...this will reduce request count to half of its previous value (will be a big plus for us) but need to take a deep look at the JSON for that...

What do you think?

**Comment by Rishabh570** [comment]
Pinging @jywarren, is it okay till now? Or do I need to do some changes...?

**Comment by jywarren** [comment]
This looks great!!! Sorry i'm slow responding. 

> I am thinking of storing the quota refill time so that we can make a fresh request every time the quota gets refilled...but need to figure out how to trigger a request dynamically...

Perhaps there is a `toolbox.fetch()` method, and you could run a `setInterval` every minute, compare the most recent timestamp fetched to the current time, and if enough time has gone by, trigger `.fetch()` again? Does the GitHub API tell us how much quota we have left in each response?

File: src/controller/controller.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+var model = require('../models/index');
```
Comment (jywarren) [inline src/controller/controller.js]: This, for now, is looking like utility methods to access the database, whereas controller code might be more like code you call for each specific use. So maybe these should be exported from something like `/src/storage/util.js`? Just a thought, i may not be right...

**Comment by Rishabh570** [comment]
> Perhaps there is a toolbox.fetch() method, and you could run a setInterval every minute, compare the most recent timestamp fetched to the current time, and if enough time has gone by, trigger .fetch() again?

Yes, that seems possible...so we will end up waiting for quota refill and not for 24hrs to pass by?

> Does the GitHub API tell us how much quota we have left in each response?

Yes, there's response header which tells about the remaining requests...that might help!

> So maybe these should be exported from something like `/src/storage/util.js`?

Hmm...this seems a suitable name :) I'll do this...

@jywarren Perhaps I'm also slow because of my exams :sweat_smile:

**Comment by jywarren** [comment]
Well, we can start by waiting a fixed amount of time, but then later
upgrade it to try to detect when the earliest opportunity is to refresh.
One step at a time will be easier to implement and this is already a big
project! Let's just make a to-do (issue?) to follow up on anything we can
split out and do later.

No rush at all - i'm busy too because of end of GCI! Please do well on your
exams and don't exhaust yourself. This will wait :-)

**Comment by Rishabh570** [comment]
> Well, we can start by waiting a fixed amount of time, but then later
upgrade it to try to detect when the earliest opportunity is to refresh.

Yes, trying to put everything in one go will create problems...

> Let's just make a to-do (issue?) to follow up on anything we can
split out and do later.

I'll surely do that, it will help me in remembering things as well ;)

> i'm busy too because of end of GCI!

Yeah, I'll try to help as much as I can in GCI part :smiley:

**Comment by Rishabh570** [comment]
@jywarren So my exams are finally over :sweat_smile: and I'll be working on it now...so couple things,

- As this is getting big, we need tests to ensure things are working the way they are supposed to...so we need to use `Jasmine` for testing purposes here...?
- As this PR is big in itself, I think I should open a follow-up PR for adding tests and just not mix-up everything here...what do you think?

The code that is present in `controller.js` should be a storage utility file so I guess I should add that code in `models` folder (with a file name `utils.js`)...

(I'll open up issues for the things we need to in later stages)

**Comment by jywarren** [comment]
I think Jasmine is a good idea. Have you seen how it's set up in other repositories? Teaspoon may also be an options... @Souravirus has recently worked on setting Teaspoon up in `plots2` maybe has opinions? Jasmine is nice though - can test interactive features on web pages as well as pure JavaScript functions. 

I think it's OK to add tests in a follow-up, esp since we don't already have them. Shall we perhaps make a list of things to test in our first run? Maybe 4-5 major functions that could give us confidence that code.publiclab.org and the embed code at for example http://code.publiclab.org/examples/embed.html#o=publiclab&r=all&l=gci-candidate ?

**Comment by jywarren** [comment]
And congrats on your exams!!!

**Comment by Rishabh570** [comment]
> Maybe 4-5 major functions that could give us confidence that code.publiclab.org and the embed code at for example http://code.publiclab.org/examples/embed.html#o=publiclab&r=all&l=gci-candidate ?

Yep, I guess the things that ensure the overall working of the site are...

- List of all contributors (this is stored in indexedDB from now)
- List of recent contributors (this uses localStorage as of now because this PR has only migrated all contributors/repo contributors part to indexedDB)
- The function of `embed.html` filter that you mentioned (the logic behind filters can be tested)
- Issues list (never touched this part)

@jywarren Is anything I'm missing?

**Comment by Rishabh570** [comment]
Yes...tests for what functions are returning would be the primary and main motive at first and how it's getting displayed can be added then...

**Comment by Rishabh570** [comment]
yeah, unit testing first :+1: 
I'm happy to have a chance to contribute here :)

**Comment by jywarren** [comment]
Awesome. This is very cool.

**Comment by Souravirus** [comment]
Hi @Rishabh570, if you can add tests through teaspoon-mocha then it's fine. Else if you want to add jasmine tests, you have to set the previous tests again to jasmine. Also, the phantoms was not testing the new JavaScript features. So, just see to it. Thank you!!

**Comment by Rishabh570** [comment]
I would work on teaspoon-mocha if this is what we implement usually in projects...

>  you have to set the previous tests again to jasmine

But I don't understand this, this project doesn't have any tests, right? I don't get _previous tests_ part...

> Also, the phantomjs was not testing the new JavaScript features.

I don't know where phantomjs is implemented...are you talking about some other project? 

@Souravirus Can you please tell me about this?

**Comment by jywarren** [comment]
Ah, apologies, Sourav - this is actually the code for
https://code.publiclab.org, not plots2, and I was interested in your input
mainly because you'd led the conversion of plots2 from jasmine to teaspoon.
A number of our projects are in Jasmine (PublicLab.Editor, for one) and it
seems pretty nice. I have less experience with Teaspoon but wanted to
suggest it. Is Jasmine still active? If so, perhaps we could stick with it
since it's in a number of other Public Lab projects.

Thanks for thinking this through with us!

**Comment by Souravirus** [comment]
Yeah sure you can test it with jasmine. I am sorry I didn't see this was community-toolbox. Actually, I was commenting with my cell phone and forgot to see the repository.

**Comment by Rishabh570** [comment]
Hi @jywarren, so I was able to figure out what was causing problem in the code...Actually, there is an event listener in `index.js` (of models) which fires `initialise()` function when DOM is loaded and it takes quite a much time (approx 500ms) to execute successfully. Also, it takes this time (of 500ms) at the very first when the user creates a fresh database in indexedDB for the first time, any page load after that doesn't take that much...so the problem was, I was firing `showRepoContributors` instantly as the page loads (which needs a database) while `initialise()` function was running (not completed creating the DB) so error was because of DB not being there when code needs it...

I've put a `setTimeout` in the `index.html` where I'm calling `shopRepoContributors` and `getAllContributors` which solves the problem (By putting setTimeout, DB gets created by the time code actually needs it)...Actually, I can't think of any other solution of this other than putting setTimout there. But I don't think 500ms would be much noticable :thinking: 

Btw here is working GIF of all contributors and repoContributors getting loaded from the DB,

mvc_1

**Comment by Rishabh570** [comment]
and here is the screenshot of how indexedDB looks,

screenshot from 2018-12-21 23-34-15

**Comment by jywarren** [comment]
oh wow! Yes, the tests will also make it easier for me to understand your
code, as it will 'demonstrate' the proper usage. This is looking awesome!
I'm catching up after the holidays but excited to see it!

**Comment by Rishabh570** [comment]
Hi @jywarren, there's a lot of work remaining here...having a little headache today so won't be able to work on it now :( But I'll get on to it as soon as I feel better...

As an aside, I saw this project on the ideas list for GSoC'19 (shifting to mvc etc.). I would like to brainstorm what other features/enhancements we would like to incorporate to this project...should I open an issue for it, as a discussion thread?
Thanks!

**Comment by jywarren** [comment]
aw, i hope you feel better soon!

Yes, that's a great idea!

**Comment by Rishabh570** [comment]
@jywarren 
**Quick update:** I've rebased MVC work on top of latest commits (finally :sweat_smile: )...working on recent contributors section now (support for showing recent contributors repository-wise).
I'll try to finish this PR's work within this week :)

**Comment by Rishabh570** [comment]
@jywarren I've pushed some modifications for now...it's live on http://rishabh570.me/community-toolbox/
I've incorporated MVC with all the features we're supporting till now...it's still in development so errors might occur...you can check if the page is showing data repo-wise now.
working on code readability now
Please take a look :)

PS: I guess we'll need a proper documentation for it in near future :memo:

**Comment by Rishabh570** [comment]
Hi @jywarren, please take a look when you find time :)
_Its still in development though, so not ready to merge_

File: src/models/index.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    dbReq.onerror = function(event) {
+        console.log('error opening database');
+    }
+}
+
+// We make sure that initialize function gets fired once the DOM content is loaded
+window.addEventListener('DOMContentLoaded', initialize());
+
+
+function saveContentToDb(queryKey, queryContent) {
```
Comment (jywarren) [inline src/models/index.js]: This is looking amazing. One way to begin making it more readable would be to break these into separate files and use `require()` where possible.

File: src/models/utils.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
+var model = require('./index');
```
Comment (jywarren) [inline src/models/utils.js]: Nice and readable!!!!

File: src/scripts/community-toolbox.js [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+                        return result;
+                      })
+            }
+          });
+        }
+      }
+    });
+  }
+
+
```
Comment (jywarren) [inline src/scripts/community-toolbox.js]: This file is starting to get a bit long; if possible, we should break it up into subfiles along MVC lines!

File: src/utils/getAllContribsUtility.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+// Fetches all the publiclab's repositories
+function getAllRepos(org) {
+
+  // This array is used to store all the repositories fetched from Github
+  let repos = [];
+
+  return fetch('https://api.github.com/users/publiclab/repos?sort=pushed&direction=desc&per_page=100')
```
Comment (jywarren) [inline src/utils/getAllContribsUtility.js]: Here, shouldn't we point it at `org` though we can default to `publiclab`?

File: src/utils/getRecentCommitsUtility.js [hunk 6]
```diff
@@ -0,0 +1,193 @@
+var getAllContribsUtility = require('../utils/getAllContribsUtility');
+var model_utils = require('../models/utils');
+
+// Fetches recent month commits for a particular repository 
+function fetchRecentMonthCommits(repo, queryTime) {
+    let commitersSet = new Set([]);
+    let result=[];
+    return fetch(`https://api.github.com/repos/publiclab/${repo}/commits?since=${queryTime}`)
```
Comment (jywarren) [inline src/utils/getRecentCommitsUtility.js]: Same here on `org` - that may also mean we need a param for `org`

File: src/utils/getRecentCommitsUtility.js [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+function fetchAllRecentMonthCommits(repos, queryTime) {
+    let results = [];
+    let commitersSet = new Set([]);
+    let timeToday = (new Date).getTime();
+
+    // We take only 10 repos just for API quota reasons
+    let splicedRepos = repos.splice(0,10);
+
+    let promises = splicedRepos.map(function mapToEachRepo(repo, i) {
+        return fetch(`https://api.github.com/repos/publiclab/${repo}/commits?since=${queryTime}`)
```
Comment (jywarren) [inline src/utils/getRecentCommitsUtility.js]: `org`

File: src/utils/getRecentCommitsUtility.js [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+                // Store recentMonthCommits and recentMonthCommitsExpiry in the database
+                console.log(`in fetchALLRecentMonthCommits, storing recent-ALL-month-commits`);
+                model_utils.setItem('recent-all-month-commits', results);
+                model_utils.setItem('recent-all-month-expiry', timeToday);
+                return results;
+            });
+}
+
+
+function getCommitsLastWeek(org, repo) {
```
Comment (jywarren) [inline src/utils/getRecentCommitsUtility.js]: Basically it's OK to have one function per file if the functions are this big and each serves a single purpose.

**Comment by Rishabh570** [comment]
@jywarren Okay! For now, I'll bump up the version number and will provide `org` parameter wherever needed. 
We can do the refactoring in a follow-up PR :)

File: src/models/index.js (hunk 2 above)
Comment (Rishabh570) [reply src/models/index.js]: Yeah, that will make it more clear and more specific.

File: src/scripts/community-toolbox.js (hunk 4 above)
Comment (Rishabh570) [reply src/scripts/community-toolbox.js]: Yes, I also feel like so...I'll try to do it :+1:

File: src/utils/getAllContribsUtility.js (hunk 5 above)
Comment (Rishabh570) [reply src/utils/getAllContribsUtility.js]: On it.

File: src/utils/getRecentCommitsUtility.js (hunk 8 above)
Comment (Rishabh570) [reply src/utils/getRecentCommitsUtility.js]: Yeah, I will try to refactor this chunk in a follow-up PR...

**Comment by Rishabh570** [comment]
Currently, I'm inspecting the deletion part of the database...for some reason, entries are not getting deleted from the database right now...
I've pushed the other required changes that you mentioned.

**Comment by jywarren** [comment]
OK, just tell me when you're ready to merge and we can do it and publish to
NPM. This is so exciting!!!

**Comment by Rishabh570** [comment]
@jywarren Yeah I'll notify you as soon as I complete it...are we going to publish it as a NPM package? That'd be super cool :smiley:

**Comment by Rishabh570** [comment]
Hi @jywarren, I've pushed my commit regarding entry deletion error from the DB...my changes are live at: http://rishabh570.me/community-toolbox/
But I've modified some DB related config (changed the name of the DB index) so you need to delete the old version of the database from the `Application` tab of chrome dev tools (in the indexedDB section) in order to test the latest version.

Also, I want to have it tested by people so that we can be sure of its working, then only we can focus on refactoring & testing support... :thought_balloon: 
What do you think?
Thanks!

**Comment by Rishabh570** [comment]
Hi @jywarren, please take a look :) It's running fine for me btw...

**Comment by jywarren** [comment]
Hi! I had a couple days off but will try to catch up tomorrow. Looking
forward to it!!! And I like the idea of testing, we can use your gh-pages
and ask people in the weekly check-in maybe?

**Comment by Rishabh570** [comment]
@jywarren Oh, sorry I didn't know...
Yes, I'll share my link with everyone! :smiley:

**Comment by Rishabh570** [comment]
Hi @jywarren, I guess this is fine, it is working properly by far...what do you think, should we proceed?

**Comment by jywarren** [comment]
To merge? I'm fine with that! Do you think the documentation is pretty good
or we could do it in follow-up? Maybe later a diagram of the MVC components
would be nice.

**Comment by Rishabh570** [comment]
@jywarren Yeah so, I guess we can do documentation related stuff in a follow-up PR...diagram of the components showing the flow control would be really good (I guess we can build our documentation around the diagram, first showing the diagram then explaining every part of it).

So, to keep things in check, our (near) future goals/checklist includes,
- [ ] Testing
- [ ] Documentation
- [ ] Notifying user about the unexpected errors
- [ ] Making API requests by keeping in check the remaining API quota
- [ ] Sorting the contributors' data according to their contributions (maybe some other filter also)

Please remind me if I'm missing something...
Also, is the order of the list correct according to our priorities (top having most priority)?

Thanks!!!

**Comment by Rishabh570** [comment]
> To merge?

@jywarren Yes, it is ready to merge! :tada:

**Comment by jywarren** [comment]
Wow, awesome! Let's move the checklist into a new issue. I'll deploy to gh-pages in a moment!

**Comment by jywarren** [comment]
I also bumped the version to `v1.0.0`!

**Pull request merged**